src/lib/net/address.c

```diff
@@ -827,29 +828,11 @@ tor_addr_is_valid(const tor_addr_t *addr, int for_listening)
     return 0;
   }
 
-  /* Only allow IPv4 0.0.0.0 for_listening. */
-  if (for_listening && addr->family == AF_INET
-      && tor_addr_to_ipv4h(addr) == 0) {
-    return 1;
-  }
-
-  /* Otherwise, the address is valid if it's not tor_addr_is_null() */
-  return !tor_addr_is_null(addr);
-}
-
-/* Is the network-order IPv4 address v4n_addr valid?
- * Checks that addr is not zero.
- * Except if for_listening is true, where IPv4 addr 0.0.0.0 is allowed. */
-int
-tor_addr_is_valid_ipv4n(uint32_t v4n_addr, int for_listening)
-{
-  /* Any IPv4 address is valid with for_listening. */
-  if (for_listening) {
-    return 1;
-  }
-
-  /* Otherwise, zero addresses are invalid. */
-  return v4n_addr != 0;
+  /* For the case of IPv4 it doesn't matter if it is in network order
+   * or host order because we are testing it against 0 for validity.
+   * If for_listening is true i.e. 1 then it doesn't matter if addr
+   * is null or not i.e. 0.0.0.0 for IPv4 and :: for IPv6. */
+  return for_listening || !tor_addr_is_null(addr);
```

Comment:
This code gives an incorrect result for Unix socket addresses, and the unspecified null address.

IPv4 0.0.0.0 and IPv6 :: are the only documented exceptions. So you can't remove the AF_INET and AF_INET6 checks.